// ask-sdk-v1adapter/dist/templateBuilders/bodyTemplate7Builder.js
"use strict";
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var BodyTemplate7Builder = /** @class */ (function () {
    function BodyTemplate7Builder() {
        this.template = { type: 'BodyTemplate7' };
    }
    BodyTemplate7Builder.prototype.setImage = function (image) {
        this.template.image = image;
        return this;
    };
    BodyTemplate7Builder.prototype.setTitle = function (title) {
        this.template.title = title;
        return this;
    };
    BodyTemplate7Builder.prototype.setToken = function (token) {
        this.template.token = token;
        return this;
    };
    BodyTemplate7Builder.prototype.setBackgroundImage = function (image) {
        this.template.backgroundImage = image;
        return this;
    };
    BodyTemplate7Builder.prototype.setBackButtonBehavior = function (backButtonBehavior) {
        this.template.backButton = backButtonBehavior;
        return this;
    };
    BodyTemplate7Builder.prototype.build = function () {
        return this.template;
    };
    return BodyTemplate7Builder;
}());
exports.BodyTemplate7Builder = BodyTemplate7Builder;
