// ask-sdk-v1adapter/dist/templateBuilders/listItemBuilder.js
"use strict";
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var textUtils_1 = require("../utils/textUtils");
var ListItemBuilder = /** @class */ (function () {
    function ListItemBuilder() {
        this.listItems = [];
    }
    ListItemBuilder.prototype.addItem = function (image, token, primaryText, secondaryText, tertiaryText) {
        var listItem = {
            token: token,
            image: image,
            textContent: textUtils_1.TextUtils.makeTextContent(primaryText, secondaryText, tertiaryText),
        };
        this.listItems.push(listItem);
        return this;
    };
    ListItemBuilder.prototype.build = function () {
        return this.listItems;
    };
    return ListItemBuilder;
}());
exports.ListItemBuilder = ListItemBuilder;
