// ask-sdk-v1adapter/dist/utils/textUtils.js
"use strict";
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var TextUtils = /** @class */ (function () {
    function TextUtils() {
    }
    TextUtils.makePlainText = function (text) {
        var plainText = {
            type: 'PlainText',
            text: text,
        };
        return plainText;
    };
    TextUtils.makeRichText = function (text) {
        var richText = {
            type: 'RichText',
            text: text,
        };
        return richText;
    };
    TextUtils.makeTextContent = function (primaryText, secondaryText, tertiaryText) {
        var textContent = {
            primaryText: primaryText,
            secondaryText: secondaryText,
            tertiaryText: tertiaryText,
        };
        return textContent;
    };
    return TextUtils;
}());
exports.TextUtils = TextUtils;
