// ask-sdk-v1adapter/dist/services/v1ApiClient.js
"use strict";
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var ask_sdk_1 = require("ask-sdk");
var V1ApiClient = /** @class */ (function () {
    function V1ApiClient() {
        this.defaultApiClient = new ask_sdk_1.DefaultApiClient();
    }
    V1ApiClient.prototype.post = function (uri, headers, body) {
        var options = this.buildOptions(uri, headers, 'POST');
        return this.dispatch(options, body);
    };
    V1ApiClient.prototype.put = function (uri, headers, body) {
        var options = this.buildOptions(uri, headers, 'PUT');
        return this.dispatch(options, body);
    };
    V1ApiClient.prototype.get = function (uri, headers) {
        var options = this.buildOptions(uri, headers, 'GET');
        return this.dispatch(options);
    };
    V1ApiClient.prototype.delete = function (uri, headers) {
        var options = this.buildOptions(uri, headers, 'DELETE');
        return this.dispatch(options);
    };
    V1ApiClient.prototype.buildOptions = function (uri, requestHeaders, requestMethod) {
        var request = {
            headers: requestHeaders,
            method: requestMethod,
            url: uri,
        };
        return request;
    };
    V1ApiClient.prototype.dispatch = function (options, body) {
        options.body = body;
        return this.defaultApiClient.invoke(options);
    };
    return V1ApiClient;
}());
exports.V1ApiClient = V1ApiClient;
